File: packages/python-sdk/src/certifieddata_payments/webhooks.py
"""Webhook signature verification for CDP webhooks."""

import hashlib
import hmac
import time


def verify_webhook_signature(
    raw_body: str | bytes,
    signature_header: str,
    timestamp_header: str,
    secret: str,
    tolerance_seconds: int = 300,
) -> bool:
    """
    Verify a CDP webhook signature.

    CDP signs webhooks using HMAC-SHA256 over the string:
        ``{timestamp}.{raw_body}``

    The ``CDP-Signature`` header has the format::
        ``t={timestamp},v1={hmac_hex}``

    :param raw_body: Raw request body (str or bytes) — before any JSON parsing.
    :param signature_header: Value of the ``CDP-Signature`` header.
    :param timestamp_header: Value of the ``CDP-Timestamp`` header.
    :param secret: Webhook endpoint secret.
    :param tolerance_seconds: Maximum age of the webhook in seconds (default 300).
    :returns: True if the signature is valid and within the timestamp tolerance.
    :raises ValueError: If the signature header format is invalid.
    """
    # Parse timestamp
    try:
        ts = int(timestamp_header)
    except (ValueError, TypeError) as exc:
        raise ValueError(f"Invalid CDP-Timestamp header: {timestamp_header!r}") from exc

    # Check timestamp tolerance
    now = int(time.time())
    if abs(now - ts) > tolerance_seconds:
        return False

    # Extract v1 signature from header
    v1_sig: str | None = None
    for part in signature_header.split(","):
        part = part.strip()
        if part.startswith("v1="):
            v1_sig = part[3:]
            break

    if not v1_sig:
        return False

    # Compute expected signature
    body_bytes = raw_body.encode("utf-8") if isinstance(raw_body, str) else raw_body
    signed_payload = f"{ts}.".encode("utf-8") + body_bytes
    expected = hmac.new(secret.encode("utf-8"), signed_payload, hashlib.sha256).hexdigest()

    # Constant-time comparison
    return hmac.compare_digest(expected, v1_sig)

File: packages/typescript-sdk/src/utils/webhooks.ts
/**
 * Webhook signature verification for CertifiedData Payments.
 *
 * Signing algorithm: HMAC-SHA256
 * Signed payload: {timestamp}.{raw_body}
 * Signature header: CDP-Signature: t={timestamp},v1={hmac_sha256}
 * Timestamp header: CDP-Timestamp: {unix_timestamp}
 * Tolerance: 300 seconds
 *
 * See test-vectors/webhook-signature/ for verified test cases.
 */

export interface VerifyWebhookSignatureParams {
  /** The raw request body string (not parsed). */
  payload: string;
  /** The CDP-Signature header value: t={timestamp},v1={hmac} */
  signature: string;
  /** The CDP-Timestamp header value (unix timestamp string). */
  timestamp: string;
  /** The webhook endpoint secret. */
  secret: string;
  /** Timestamp tolerance in seconds. Defaults to 300. */
  toleranceSeconds?: number;
}

export interface VerifyWebhookSignatureResult {
  valid: boolean;
  reason?: string;
}

/**
 * Verify a CDP webhook signature.
 *
 * @example
 * const result = verifyWebhookSignature({
 *   payload: rawBody,
 *   signature: req.headers["cdp-signature"],
 *   timestamp: req.headers["cdp-timestamp"],
 *   secret: process.env.CDP_WEBHOOK_SECRET,
 * });
 * if (!result.valid) throw new Error("Invalid signature");
 */
export async function verifyWebhookSignature(
  params: VerifyWebhookSignatureParams
): Promise<VerifyWebhookSignatureResult> {
  const { payload, signature, timestamp, secret, toleranceSeconds = 300 } = params;

  // Check timestamp tolerance
  const ts = parseInt(timestamp, 10);
  if (isNaN(ts)) {
    return { valid: false, reason: "Invalid timestamp" };
  }

  const now = Math.floor(Date.now() / 1000);
  if (Math.abs(now - ts) > toleranceSeconds) {
    return { valid: false, reason: "Timestamp outside tolerance window" };
  }

  // Parse signature header: t={timestamp},v1={hmac}
  const v1Match = signature.match(/v1=([a-f0-9]+)/);
  if (!v1Match?.[1]) {
    return { valid: false, reason: "No v1 signature found in header" };
  }

  const expectedHex = v1Match[1];

  // Compute expected signature
  const signedPayload = `${timestamp}.${payload}`;
  const encoder = new TextEncoder();
  const keyData = encoder.encode(secret);
  const messageData = encoder.encode(signedPayload);

  // Use globalThis.crypto (Web Crypto API, available in Node.js ≥ 18)
  const subtle = globalThis.crypto.subtle;

  const cryptoKey = await subtle.importKey(
    "raw",
    keyData,
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"]
  );

  const signatureBuffer = await subtle.sign("HMAC", cryptoKey, messageData);
  const computedHex = Array.from(new Uint8Array(signatureBuffer))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");

  // Constant-time comparison
  if (computedHex.length !== expectedHex.length) {
    return { valid: false, reason: "Signature mismatch" };
  }

  let mismatch = 0;
  for (let i = 0; i < computedHex.length; i++) {
    mismatch |= computedHex.charCodeAt(i) ^ expectedHex.charCodeAt(i);
  }

  if (mismatch !== 0) {
    return { valid: false, reason: "Signature mismatch" };
  }

  return { valid: true };
}
